backend/src/apiserver/resource/resource_manager.go

```diff
@@ -142,6 +142,12 @@ func (r *ResourceManager) DeletePipeline(pipelineId string) error {
 	// Delete pipeline file and DB entry.
 	// Not fail the request if this step failed. A background run will do the cleanup.
 	// https://github.com/kubeflow/pipelines/issues/388
+	// TODO(jingzhang36): For now (before exposing version API), we have only 1
+	// file with both pipeline and version pointing to it;  so it is ok to do
+	// the deletion as follows. After exposing version API, we can have multiple
+	// versions and hence multiple files, and we shall improve performance by
+	// put all files of a single pipeline in one directory; and remove the whole
```

Comment:
GCS doesn't support deleting a folder. It will just batch delete every file in the virtual folder which is still as slow. We might need to implement async api for performance reason.

Reply:
Sure. I add to the comment the async option.